### PR TITLE
More tests for new request based statement distribution

### PR DESCRIPTION
**Pull request by eskimor**
Also preserve ordering of peers sending statements (so the right peer will be rewarded with reputation).

### CODE REVIEW

File: node/network/statement-distribution/src/lib.rs [hunk 1]
```diff
@@ -508,7 +508,7 @@ enum LargeStatementStatus {
 struct FetchingInfo {
 	/// All peers that send us a `LargeStatement` or a `Valid` statement for the given
 	/// `CandidateHash`, together with their originally sent messages.
-	available_peers: HashMap<PeerId, Vec<protocol_v1::StatementDistributionMessage>>,
+	available_peers: IndexMap<PeerId, Vec<protocol_v1::StatementDistributionMessage>>,
```
Comment (ordian) [inline node/network/statement-distribution/src/lib.rs:511]: my first though was it is used to ensure determinism in tests, is that true?

**Review by ordian — APPROVED** [review]
More tests are always welcome :)

File: node/network/statement-distribution/src/lib.rs (hunk 1 above)
Comment (eskimor) [reply node/network/statement-distribution/src/lib.rs:511]: no, we iterate over that map in the end to finally process arrived messages. By using `IndexMap` we ensure the messages will be forwarded in the order they were received, which is desirable as we reward the first peer more than later peers.

Comment (ordian) [reply node/network/statement-distribution/src/lib.rs:511]: I see, would you mind adding this as doc comments then?

File: node/network/statement-distribution/src/lib.rs [hunk 2]
```diff
@@ -508,7 +508,10 @@ enum LargeStatementStatus {
 struct FetchingInfo {
 	/// All peers that send us a `LargeStatement` or a `Valid` statement for the given
 	/// `CandidateHash`, together with their originally sent messages.
-	available_peers: HashMap<PeerId, Vec<protocol_v1::StatementDistributionMessage>>,
+	///
+	/// We use an `IndexMap` here to preserve the ordering of peers sending us messages. This is
+	/// desirable because we reward first sending peers with reputation.
```
Comment (bkchr) [inline node/network/statement-distribution/src/lib.rs:513]: Where is this needed?

File: node/network/statement-distribution/src/lib.rs [hunk 3]
```diff
@@ -1061,10 +1064,10 @@ async fn retrieve_statement_from_message<'a>(
 					let is_large_statement = message.is_large_statement();
 
 					match info.available_peers.entry(peer) {
-						Entry::Occupied(mut occupied) => {
+						IEntry::Occupied(mut occupied) => {
 							occupied.get_mut().push(message);
 						}
-						Entry::Vacant(vacant) => {
+						IEntry::Vacant(vacant) => {
```
Comment (bkchr) [inline node/network/statement-distribution/src/lib.rs:1070]: Line: 1078 

`if let Some(sender) = std::mem::take(&mut info.peer_sender)`

should be

`
if let Some(sender) = info.peer_sender.take() {
`

File: node/network/statement-distribution/src/lib.rs (hunk 2 above)
Comment (eskimor) [reply node/network/statement-distribution/src/lib.rs:513]: [here](https://github.com/paritytech/polkadot/blob/eae1054f06d0489f2d4dc33f04f263292ca8694d/node/network/statement-distribution/src/lib.rs#L1369)

Comment (bkchr) [reply node/network/statement-distribution/src/lib.rs:513]: Doesn't work

Comment (eskimor) [reply node/network/statement-distribution/src/lib.rs:513]: Sorry, link just became live with the push.

Comment (bkchr) [reply node/network/statement-distribution/src/lib.rs:513]: Ahh, now I see how this works :P 

I had found the position before where this reputation change was done, but I didn't followed the full call chain to see how this is connected to `available_peers`.

File: node/network/statement-distribution/src/lib.rs [hunk 4]
```diff
@@ -1061,18 +1064,18 @@ async fn retrieve_statement_from_message<'a>(
 					let is_large_statement = message.is_large_statement();
 
 					match info.available_peers.entry(peer) {
```
Comment (bkchr) [inline node/network/statement-distribution/src/lib.rs:1066]: Sorry.

Another one.  :D 

```suggestion
					let is_new_peer = match info.available_peers.entry(peer) {
```

return `true` or `false` below and remove the check above. It doesn't really make sense to access the map twice.

File: node/network/statement-distribution/src/lib.rs (hunk 2 above)
Comment (eskimor) [reply node/network/statement-distribution/src/lib.rs:513]: Sorry had to rush that req/resp PR quite a bit, so I am afraid I did not exactly make it more readable. :-S

File: node/network/statement-distribution/src/lib.rs (hunk 4 above)
Comment (eskimor) [reply node/network/statement-distribution/src/lib.rs:1066]: Fair, just found it slightly more readable this way, but in hindsight it probably isn't :-)

**Pull request merged**